mhkit/river/io/d3d.py

```diff
@@ -0,0 +1,444 @@
+    '''  
+    
+    assert isinstance(time_index, int), 'time_index  must be a int'
+    assert type(data)== netCDF4._netCDF4.Dataset, 'data must be nerCDF4 object'
+    assert variable in data.variables.keys(), 'varaiable not reconized'
+
+    max_time_index = len(data.variables[variable][:])
+    assert time_index <= max_time_index, (f'time_index must be less than the'
+                                          +'max time index {max_time_index}')
+    assert time_index >= -1, 'time_index must be greater than or equal to -1'
+
+    cords_to_layers= {'laydim': data.variables['LayCoord_cc'][:],
+                       'wdim': data.variables['LayCoord_w'][:]}
+    lay_element= 2 
+    layer_dim =  [v.name for v in data[variable].get_dims()][lay_element]
+    
+    try:    
+        cord_sys= cords_to_layers[layer_dim]
+    except: 
+        raise Exception('coordinates not recognized')
+    else: 
+        Layer_percentages= np.ma.getdata(cord_sys, False) 
+        
+    bottom_depth=np.ma.getdata(data.variables['waterdepth'][time_index, :], False)
+    if layer_dim == 'wdim': 
+        #interpolate 
+        coords = str(data.variables['waterdepth'].coordinates).split()
+        x_laydim=np.ma.getdata(data.variables[coords[0]][:], False) 
+        y_laydim=np.ma.getdata(data.variables[coords[1]][:], False)
+        points_laydim = np.array([ [x, y] for x, y in zip(x_laydim, y_laydim)])
+        
+        coords_request = str(data.variables[variable].coordinates).split()
+        x_wdim=np.ma.getdata(data.variables[coords_request[0]][:], False) 
+        y_wdim=np.ma.getdata(data.variables[coords_request[1]][:], False)
+        points_wdim=np.array([ [x, y] for x, y in zip(x_wdim, y_wdim)])
+        
+        bottom_depth_wdim = interp.griddata(points_laydim, bottom_depth,
+                                            points_wdim)
+        
+        idx= np.where(np.isnan(bottom_depth_wdim))
+        
+        for i in idx: 
+            bottom_depth_wdim[i]= interp.griddata(points_laydim, bottom_depth,
+                                             points_wdim[i], method='nearest')
+        
+    x_all=[]
+    y_all=[]
+    z_all=[]
+    v_all=[]
+    
+    N_layers = range(len(Layer_percentages))
+    for layer in N_layers:
+        layer_data= get_layer_data(data, variable, layer, time_index)
+        if layer_dim == 'wdim': 
+            z = [bottom_depth_wdim*Layer_percentages[layer]]
+        else: 
+            z = [bottom_depth*Layer_percentages[layer]]
+        x_all=np.append(x_all, layer_data.x)
+        y_all=np.append(y_all, layer_data.y)
+        z_all=np.append(z_all, z)
+        v_all=np.append(v_all, layer_data.v)
+    
+    known_points = np.array([ [x, y, z, v] for x, y, z, v in zip(x_all, y_all, 
+                                                                z_all, v_all)])
+    
+    all_data= pd.DataFrame(known_points, columns=['x','y','z',f'{variable}'])
+    
+    return all_data
+
+
+def unorm(x, y ,z):
```

Comment:
@ssolson do you think this function would be good to have in the utils module? Its very general and could be useful elsewhere...

Reply:
Agree. @browniea please move this to the utils module.

Reply:
@browniea looks like this is the last comment to address. Afterwards I think this will be good to merge

Reply:
Added to utils @ssolson did you want me to remove it from this module?